### ISSUE TITLE
Why do we need to divide the loss in case of multi-dimensional labels?

### ISSUE THREAD

**Issue by Yannik1337** [tensorflow/tensorflow#48586]
I am following TensorFlow's tutorial on [custom training][1].
There are two things that I do not understand:

First, when defining the losses, the authors point to dividing by the shape of one example, in the case of a multi-dimensional labels:

"If `labels` is multi-dimensional, then average the `per_example_loss` across the number of elements in each sample. For example, if the shape of predictions is `(batch_size, H, W, n_classes)` and labels is `(batch_size, H, W)`, you will need to update per_example_loss like: `per_example_loss /= tf.cast(tf.reduce_prod(tf.shape(labels)[1:]), tf.float32)`"

Why do we need to rescale the loss this way? And does this also hold for Cycle-GANs?

Secondly, a general question: I noticed that several functions use the `with strategy.scope()` annotation. If this object is available globally, then we can define a function on the global level like this
```
with strategy.scope():
  def example_function():
    #do stuff
```
If the object is not available globally, we have to structure the code like this:
```
def train():
  strategy = ...

  with strategy.scope():
    def sub_function1():
      #do stuff
```

This leads to one large function, `train()`, with many short sub functions. I'd like to define the sub functions on the same level as `train()`, without using a global `strategy` object. 
Is there a workaround to achieve this?

  [1]: https://www.tensorflow.org/tutorials/distribute/custom_training#define_the_loss_function

**Comment by nikitamaia** [comment]
Hi @Yannik1337, for multi-dimensional labels, you still need to divide the loss by the global batch size later with `tf.reduce_sum(per_example_loss) / GLOBAL_BATCH_SIZE`
```
per_example_loss = loss_function(labels, predictions) # shape: [batch_size, H, W]
per_example_loss /= tf.reduce_prod(labels.shape[1:])
loss = tf.reduce_sum(per_example_loss) / GLOBAL_BATCH_SIZE
```

which is equivalent to if you're using a single replica
```
per_example_loss = loss_function(labels, predictions)
loss = tf.reduce_mean(per_example_loss)
```

I don't think it's completely obvious from the tutorial that you still need to divide by the global batch size (like in the case where your labels are not multi-dimensional), so i hope this clears things up!

Also, you only need to wrap the creation of your model variables (including metrics and optimizers) within `strategy.scope`. You do not need to put your training code within the scope of the strategy. 

Github is for bugs and performance issues, so I am closing this issue now. Stack Overflow is a better place to go if you have support questions since there's a larger community there to respond. Thanks!
